#462 Add "CI/CD Status" column

keydepth opened the pull request:
Issue Number / Link: #462

Summary of the issue:
Implemented the "CI/DI Status" column feature

Description outlining how this pull request resolves the issue:
- Add "CI/DI Status" column and column settings.
- Add configuration and configuration menu/dialog of Provider, GitURL and Access Token.
- Add Liblary for HTTPS request.
- Add test code.

"CI/DI Status" column 
![image](https://user-images.githubusercontent.com/35625044/110242451-91562d80-7f99-11eb-9bde-b5d04c604f8a.png)

Configuration Menu
![image](https://user-images.githubusercontent.com/35625044/110242421-623fbc00-7f99-11eb-8ef9-5433d3995777.png)

Configuration Dialog
![image](https://user-images.githubusercontent.com/35625044/110242427-6cfa5100-7f99-11eb-883a-bdf4ba72a7ed.png)

keydepth committed:
#462 Pipeline/Workflow status for GitLab/GitHub

keydepth committed:
#462 Removed CustomPipeline.

keydepth committed:
#462 Renamed "Pipeline Status" to "CI/CD Status"

keydepth committed:
#462 Updated frontend behavior of "CI/CD Status"

keydepth committed:
#462 Updated test code for CI/DI Status.

keydepth committed:
#462 Updated library in package.json

keydepth committed:
#462 Updated GitLab Authorization of CI/DI Status

mhutchie commented:
Thanks @keydepth for raising this pull request!

I’ll review this PR in the next few days - I’m just  finishing the implementation and testing of a feature that I’ve committed to deliver in Git Graph 1.30.0, and need to get a beta release out for it ASAP.

keydepth committed:
#462 Fiexd typo CI/DI to CI/CD

keydepth commented:
I found a lot of typo and fixed typo "CI/**DI**" to "CI/**CD**".

commit 64ec2cd7c7ded131ea942952c9cd49cfa0f3211b

"CI/CD Status" column
![image](https://user-images.githubusercontent.com/35625044/110260908-6d263b00-7ff1-11eb-9592-b7efc78c1f1c.png)

Configuration Menu
![image](https://user-images.githubusercontent.com/35625044/110260925-7b745700-7ff1-11eb-8e73-51d02f3a54d8.png)

Configuration Dialog
![image](https://user-images.githubusercontent.com/35625044/110260945-9941bc00-7ff1-11eb-9c57-03f044cd6a61.png)

keydepth committed:
Merge branch 'develop' into develop

keydepth committed:
#462 Fiexd typo LF to CRLF

keydepth committed:
#462 Fiexd typo LF to CRLF

keydepth commented:
@mhutchie 
I'v fixed conflict of src/dataSource.ts in this PR. ＼(^o^)／

mhutchie reviewed (changes requested):
Hi @keydepth,

Firstly, thank you for raising the pull request! It's great to have a functional implementation of this feature request.

There are a number of changes required before this functionality is production ready, please address them:
1. `DataSource.getCICDs` blocks the loading of the Git Graph View until all API requests (relatively slow) are made. The Git Graph View must always be responsive, so `DataSource.getRepoInfo` & `DataSource.getCommits` are extremely streamlined to make the commits load as quickly as possible. Anything used to enrich the commits, that can be done later, should be done later. For example, avatars are only fetched after the commits have loaded - this ensures commits always load first and can be interacted with as quickly as possible, and the avatars are loaded afterwards (without preventing the user from doing anything). This is definitely the case for this feature.
    * Even in a repository with ~300 builds, the view's load time has increased by over four seconds by having to make three API requests. As a large portion of Git Graph users work in repositories with thousands of commits & builds, the current implementation is simply not feasible. Additionally, by using `Promise.all` to load all subsequent pages of data simultaneously, if the number of pages was large I'd expect the API's to start blocking the requests due to them being overloaded.
2. Every time the Git Graph View refreshes, it has to re-fetch all of the builds via the API's. This is quite wasteful, so it would be best if a caching mechanism was implemented to reduce the number of calls made to the API's.
3. The `DataSource` class is meant to focus on running Git commands and parsing their output. It would be best if `DataSource.getCICDs` was moved into it's own class (like `./src/avatarManager.ts`), and the DataSource just requests CI/CD information from it.
4. The worded statuses (e.g. "success", "failure") should likely be replaced with coloured `SVG_ICONS` (e.g. `SVG_ICONS.passed` & `SVG_ICONS.failed`). This would make it easier for users to spot a red/failed build, out of all of the green/passed builds.
5. Many users have expressed that they want as much space to be available for the Graph & Description columns as possible (especially those who work on laptops with lower screen resolutions). I think many of these users would find the current additional column too wide. I can think of two options to resolve this (but feel free to try out some of your own):
     * Use a shorter column name (e.g. "CI/CD") and the `SVG_ICONS` in (4), to reduce the column width.
     * Move the `SVG_ICONS` in (4) to the end of the Description column text (similar to how GitHub does it). This removes the need to have any additional column.
6. For such a complex piece of code with API integrations, it would be useful to have more logging. This would allow users to see when API calls are made, and more information if they fail.
7. I try to keep the extension as lean as possible. By including the `request` and `request-promise` dependencies, they transitively added 48 new dependencies of the extension (almost 5.5Mb in size) - making the extension over 6.5 times bigger than it's current size. While I wouldn't mind this if the dependencies offered significant value, all you're using them for is to simplify making HTTPS GET requests. In contrast, this could equally be achieved by creating a `./src/utils.ts` method that wraps an HTTPS GET request in a promise in roughly 20 lines of code (this is definitely my preference).

keydepth commented:
@mhutchie 
I really appreciate your review.
I understood your point of view.

I want you just idea of (6)How to notify API error to user. 
I think vscode.window.showInformationMessage is reasonable way so far. 

I'll update this code. (^o^)

keydepth committed:
#462 Added cicdManager and changed relation files

mhutchie commented:
Hi @keydepth,

Thanks for your response, and that you understand my point of view! I really appreciate that you want to work on this feature request, have already implemented a great and fully-functional solution, and are willing to make some code improvements to refine the solution.

For (6):
* When an error occurs due to the user CI/CD config (e.g. failed to authenticate), you can use `vscode.window.showErrorMessage` so the user is made aware that they need to update their configuration. I have a simple `showErrorMessage` wrapper method for it exported from `./src/utils.ts` that should be used.
* Alternatively, for any general logging of when API requests are made, you can just use the Logger classes `Logger.log` method. This sends the logs to the extension’s Visual Studio Code Output Channel.

keydepth commented:
Hello! @mhutchie 

Thank you for your advice.
Now, I'm implementing ```./src/cicdManager.ts``` based from ```./src/avatarManager.ts```.
I need several day to update this feature. Please wait... 😀

keydepth committed:
#462 Updated cicdManager and reset DataSource.

keydepth committed:
Merge commit '245863645af' into develop

keydepth committed:
#462 Updated CI/CD style

keydepth committed:
#462 Updated caching timeing of CI/CD

keydepth committed:
#462 Updated caching timeing of CI/CD

keydepth commented:
Hello! @mhutchie 

I'v updated CI/CD status as below.
Pleease review this pull request.

(1) ```DataSource.getCICDs``` was removed and ```./src/cicdManager.ts``` was created.
(2) Added caching mechanism. When initial, all page will load, for 10 minutes top page will load, after 10 minutes or user config change all page will load.
(3) ```DataSourc```e was restored.
(4) Added SVG_ICONS (SVG_ICONS.passed, SVG_ICONS.failed & SVG_ICONS.inconclusive).
(5) Use a shorter column name "CI/CD"
(6) Added log of API status.
(7) Replaced ```request``` and ```request-promise``` to ```https```


![image](https://user-images.githubusercontent.com/35625044/111082860-d2c47b00-854d-11eb-8757-5a509639f375.png)

keydepth committed:
#462 Updated GitHub status of CI/CD

keydepth commented:
Bug fixed for status of GitHub when "conclusion" is null (like "status" is in progress).

![image](https://user-images.githubusercontent.com/35625044/111083334-6f881800-8550-11eb-85e1-3d3fe9ba3331.png)

mhutchie commented:
Thank you for making all those changes, it's really appreciated!

This weekend I completed the development of features for Git Graph 1.30.0, and I'm now shifting to reviewing and testing it (targeting 21/03/2021 for it's release).

As soon as I have some time this week (in between preparing the release), or after the release on 21/03/2021 (worst case), I'll review your changes.

keydepth committed:
#462 Updated method of refresh CI/CD status

keydepth committed:
#462 Fiexd package.json typo CRLF to LF

keydepth committed:
#462 Fiexd package.json typo CRLF to LF

keydepth committed:
#462 Fixed method of table data in CI/CD status

keydepth committed:
#462 Updated CI/CD status
Added CI/CD datail for GitHub
Added multiple CI/CD view
Added debug message of API
Added maximum page for CD/CD fetch

keydepth committed:
Added Jenkins, multiple job, tooltip for CI/CD

keydepth committed:
Merge commit '31806caa' into develop

keydepth committed:
#462 Fixed Lint check

keydepth committed:
#462 Updated help message

keydepth commented:
Hello! @mhutchie 

I'v updated CI/CD as below for useful CI/CD.
Please revew this update.

1. Added multiple CI/CD (deleted status string)
2. Added tooltip on icon. (Status string was deletd, we need some information)
3. Added CI/CD datail for GitHub/GitLab (There are external job)
4. Added debug message of API (Visual Studio Code Output Channel)
5. Added maximum page configuration for CD/CD fetch
6. Added Jenkins CI/CD (Jenkins V2 Pipeline checkout-GitSCM )

multiple CI/CD and tooltip in column
![image](https://user-images.githubusercontent.com/35625044/112728316-f2a26880-8f69-11eb-9498-cc38cc6a60fd.png)

multiple CI/CD and tooltip  in detail
![image](https://user-images.githubusercontent.com/35625044/112728382-5af14a00-8f6a-11eb-8c36-6eafeab681b0.png)

maximum page configuration
![image](https://user-images.githubusercontent.com/35625044/112728468-c9cea300-8f6a-11eb-9d7b-0ecb3a524e30.png)

Jenkins configuration
![image](https://user-images.githubusercontent.com/35625044/112728523-0ef2d500-8f6b-11eb-8733-e3d7a8c00933.png)

keydepth committed:
#462 Bug fixed opacity of CI/CD column

keydepth commented:
Bug fixed opacity of CI/CD column.

Before fix
![image](https://user-images.githubusercontent.com/35625044/112745586-83b62580-8fe4-11eb-872f-1f90c3007bee.png)

After fix
![image](https://user-images.githubusercontent.com/35625044/112745806-05f31980-8fe6-11eb-95c2-f4122747d77e.png)

mhutchie reviewed (changes requested):
Hi @keydepth,

Thank you for making all those changes, you've made some great progress!

Here are some things I noticed, that either need to be resolved, or need your response / thoughts:
1. The commit author avatars are no longer in the correct location (it should be on the top right):
![image](https://user-images.githubusercontent.com/5315753/113543614-1466b400-962a-11eb-9ef8-4d4384ee3dea.png)
2. Please try out the view with the the extension's repository, and scroll back to before 22/11/2019 (prior to this data I used Azure Pipelines). There are some weird behaviours I've observed with the CI/CD column - I'd assume these are because only GitHub Action statuses are currently being fetched when the view is loaded.
3. All caching done by `CicdManager` should be done per repository in the workspace (e.g. `this.cicds[repo][hash]`) - this is also needed to handle the unlikely (but theoretically possible) scenario that the same commit hash occurs in multiple repositories.
4. It doesn't look like `CicdManager.emitCICD` can ever reject, so all the associated rejection handling code can likely be removed.
5. As you're using access tokens to authenticate with the API's, they should be encrypted before being stored in the Visual Studio Code Workspace Momento, and decrypted upon retrieval.
6. Communication between the front-end and back-end has a performance penalty (compared to internally within the back-end), so any unnecessary data transfer should be avoided. For example, `RequestFetchCICD` doesn't need to include `cicdConfigs` in each message, as all repository state (including `cicdConfigs`) can be retrieved from the `RepoManager`.
7. As you're adding the extension setting's `git-graph.repository.commits.fetchCICDs` & `git-graph.repository.commits.fetchCICDsPage`, you don't need to have `git-graph.fetchCICDs` & `git-graph.fetchCICDsPage`. I only use this pattern when I rename an existing extension setting, and want to have backwards compatibility with the depreciated setting.
8. Currently you have the CI/CD column visibility determined by the extension setting `git-graph.defaultColumnVisibility` and the repository state. I think the column visibility should just be determined by checking whether any `cicdConfigs` are defined for the repository. This way, the column is only shown when users have configured CI/CD for the repository.
9. I don't think the extension setting `git-graph.repository.commits.fetchCICDs` is necessary, as the user is already opting-in by configuring the CI/CD Status Configuration for a repository. Additionally, it appears that even if the user has disabled `git-graph.repository.commits.fetchCICDs`, the API's are still being called.
10. Having the extension setting `git-graph.repository.commits.fetchCICDsPage` is great idea, however as there's currently no mention of how many items are in a "page", there's no easy way for users to know what the value they set actually means. I think it would probably be best to rework this setting as `git-graph.repository.commits.fetchCICDsMaximumStatuses` (default: 1000), and calculate the number of pages by `Math.ceil(fetchCICDsMaximumStatuses / per_page)`.
11. `GitGraphView.requestCICDs` currently evaluates the two if conditions for every commit that's loaded in the view, when they can just be evaluated once, if they're moved outside of `this.commits.forEach(...)`.

Before you reply with more changes for me to review, please do an extra check over all of your code to make sure you think it's in a production-ready state. To review such a large PR takes hours of my time, which is time I'm not able to spend implementing features raised by others (that offer more net value for all users), and don't have the complexities / hesitations I raised in [this comment](https://github.com/mhutchie/vscode-git-graph/issues/462#issuecomment-777075653).

keydepth committed:
#462 Restored position of commit author avatars

keydepth committed:
#462 Changed API message

keydepth committed:
#462 Removed rejection handling code of emitCICD

keydepth committed:
#462 Removed git-graph.fetchCICDs and Page

keydepth committed:
#462 Removed commits.fetchCICDs

keydepth committed:
#462 Changed fetchCICDsPage to MaximumStatuses

keydepth committed:
#462 Updated evaluates of requestCICDs

keydepth committed:
#462 Fixed compare of object

keydepth committed:
#462 Updated evaluates of requestCICDs

keydepth committed:
#462 Changed data of CicdManager per repository

keydepth committed:
#462 Updated CI/CD column visibility

keydepth committed:
#462 Changed data of CicdManager per repository

keydepth committed:
#462 Reduced cicdConfigs  from RequestFetchCICD

keydepth commented:
Hello! @mhutchie 

Thank you for your review and great comment.
Now I'm goint to implement the feature.
And I'll check the code I wrote to production-ready state. :)

I want you the idea of (5) tokens encryption,
My idea is below...
1. Create and save internal password in Momento by ```getNonce()```.
2. Encrypto and Decrypto by ```crypto.createCipheriv()```.
 like this https://gist.github.com/vlucas/2bd40f62d20c1d49237a109d491974eb

Is this method acceptable?

mhutchie commented:
Hi @keydepth,

Yes, using the Node.js `crypto` library for encryption / decryption will definitely provide suitable security for user credentials.

keydepth committed:
#462 Updated CI/CD column visibility

keydepth committed:
#462 Reduced cicdConfigs  from RequestFetchCICD

keydepth committed:
#462 Token encryption / decryption

keydepth committed:
#462 Updated CI/CD column visibility

keydepth committed:
Merge commit 'b9112e60f5fb3d' into develop

keydepth commented:
Hello! @mhutchie 
Now I’m creating testcase for ```cicdManager.ts```
It's like a ```avatarManager.test.ts```.

> I'd assume these are because only GitHub Action statuses are currently being fetched when the view is loaded.

I used GitHub API "/repos/{owner}/{repo}/actions/runs". This API can take lists all workflow runs for a repository. It seems a reasnable API to get action results for reduce API call. 
But this API couldn't get external result of CI/CD. (This time is Azure Pilelines)
https://docs.github.com/en/rest/reference/actions#list-workflow-runs-for-a-repository


To get external result of CI/CD, I found another GitHub API "/repos/{owner}/{repo}/commits/{ref}/check-runs".
This API can take list check runs for a Git reference (by each commit hash). And can get external result of CI/CD. But it's not reasnable way to get each commit action results. (it's need too many API calls.)
https://docs.github.com/en/rest/reference/checks#list-check-runs-for-a-git-reference


So, I used actions/runs in overview table and check-runs in detail view.
Is this solution acceptable?

mhutchie commented:
While I understand the limitations you’ve found so far with the GitHub API’s, from a users perspective the inconsistency will almost certainly lead to confusion, or worse still, mislead them into thinking that the all builds were successful, when it’s possible some builds failed (but they just aren’t shown in the CI/CD column).

If there just isn’t a way to work around this (with the GitHub API’s, or others), I’d almost be tempted to only show the build statuses in the Commit Details View (and not in a CI/CD column). This eliminates any confusion, and significantly simplifies the implementation.

keydepth commented:
Hello. @mhutchie 

I understood your point of view. I think this limitation is not acceptable. User mislead is reduce quality of this extention.

In this time, I will change this PR only show the build statuses in the Commit Details View.
In the future, if we found the way to work around this (with the GitHub API’s, or others), I'd like to add CI/CD column. 😀

keydepth committed:
#462 Updated CI/CD timeout

keydepth committed:
Merge commit '5681ac724bea4a' into develop

keydepth committed:
#462 Removed CI/CD column

keydepth committed:
#462 Updated GitLab detail url

keydepth committed:
#462 Removed CI/CD column

keydepth committed:
#462 Added allow_failure style for GitLab

keydepth committed:
#462 Added allow_failure tooltips for GitLab

keydepth committed:
#462 Added comment of Encrypt/Decrypt

keydepth committed:
#462 Updated error check for CICD

keydepth committed:
#462 Added test case for CICD

keydepth committed:
#462 Removed rejejct for emitCICD

keydepth commented:
Hello. @mhutchie

I've implemented CD/CD.
Pleease review this pull request.

1. Corrected Avatars location
2. All caching done per repository/hash
3. Removed code of `CicdManager.emitCICD` rejection handling  
4. Token encryption/decryption
5. Updated communication between the front-end and back-end
6. Removed `git-graph.repository.commits.fetchCICDs`, `git-graph.fetchCICDs` & `git-graph.fetchCICDsPage`.
7. Renamed `git-graph.repository.commits.fetchCICDsPage` to `git-graph.repository.commits.fetchCICDsMaximumStatuses` and changed calculation the number of pages
8. Removed CI/CD column, only show the CI/CD statuses in the Commit Details View.
9. Added test case for CI/CD

keydepth committed:
#462 Updated test case of CI/CD date

keydepth commented:
Hello. @mhutchie

Oops. I have mistake the date format of test case.
My envirnment is GMT+9, but GitHub environment is UTC(may be...).
So, Test case date format is different.
I've updated this expected value of test case.

My envirnment is ...
`Wed Apr 22 2020 21:40:58 GMT+0900 (GMT+09:00)`

GitHub environment is ...
`Wed Apr 22 2020 12:40:58 GMT+0000 (Coordinated Universal Time)`

Test result is passed 100% in my envirnment.
![image](https://user-images.githubusercontent.com/35625044/119260995-3cc56480-bc10-11eb-81fc-037f715d9254.png)

keydepth committed:
#462 Updated test case of CI/CD date for LINT

keydepth commented:
Hello. @mhutchie

Oops, again. I have mistaken the LINT of test case.
I've updated this expected value of test case.

mhutchie commented:
Hi @keydepth,

For a number of reasons outside of my control I've been unable to work on the extension much the last few weeks. I'm currently just finishing the last features for Git Graph 1.31.0, which I then need to test and release as soon as possible. I'll review your changes as soon as Git Graph 1.31.0 is released. Apologies for the delay.

keydepth commented:
Hello! @mhutchie 
How are you?

If you have time, please review this PR.
I'd like to use this extention option. (^^/
